### PR TITLE
transforms: add Random Erasing for image augmentation

### PR DESCRIPTION
**Pull request by zhunzhong07**
Random Erasing randomly selects a rectangle region in an image and erases its pixels with random values. It can reduce the risk of overfitting, and improves CNN baselines in image classification, object detection and person reidentification.

I found that this augmentation method has been widely used in image classification (CIFA-10, CIFAR-100) and person re-identification. 

Also, it could achieve improvements on ImageNet: +0.7% in Prec@1 for ResNet-50, +0.33% in Prec@1 for ResNet-34.

Therefore, I think it would be valuable to users.

'Random Erasing Data Augmentation' by Zhong et.al. https://arxiv.org/pdf/1708.04896.pdf

A parallel work is "Improved Regularization of Convolutional Neural Networks with Cutout" proposed by  DeVries. https://arxiv.org/pdf/1708.04552.pdf

Previous pull request and issues #335 #226 #420

### CODE REVIEW

**Comment by zhunzhong07** [comment]
@alykhantejani Could you pay attention to this method and advice the member to add it into the transform. I think random erasing is a useful augmentation method that will often be used in vision tasks.

**Comment by ekagra-ranjan** [comment]
This transform can be handy during self-supervision training.

**Comment by rwightman** [comment]
I've used random erasing successfully in metric loss (triplet) training  and larger image (224x224+ imagenet like) training with good results. I think it would be a worthwhile addition. 

@zhunzhong07 .. I found that the per-pixel version was quite useful for the problems but you didn't include it in your github  impl or this PR? In my experiments, normally distributed pixels, after image normalization worked well. Uniform dist caused convergence issues later in training. I perform the RE operation once the tensors are on the GPU as part of a GPU prefetching loader/collate/normalize. 

Feel free to copy any ideas: https://github.com/rwightman/pytorch-image-models/blob/master/data/random_erasing.py

**Comment by zhunzhong07** [comment]
@rwightman Thanks for your advice. In PR, I have included per-pixel mode by `v = torch.rand(img.size()[0], h, w)`. I have checked your code and notice that the per-pixel value should be normally distributed. I will fix this bug. Thank you! 

One request. I don't have enough GPUs to train a model on imagenet right now. So, if you already have the results, could you also provide some results of training w/ or w/o random erasing on imagenet?

**Comment by rwightman** [comment]
@zhunzhong07 I'll run some imagenet trainings to support this. I don't think I have two historical runs with all the hyper-params and results recorded that didn't have some sort of change in library versions, other hyper params, machines etc... I'll let you know how it goes

File: test/test_transforms.py [hunk 1]
```diff
@@ -1317,6 +1317,23 @@ def test_random_grayscale(self):
         # Checking if RandomGrayscale can be printed as string
         trans3.__repr__()
 
+    def test_random_erasing(self):
```
Comment (ekagra-ranjan) [inline test/test_transforms.py]: Can you make the test a bit more stronger by checking if the region around the erased patch is equal to the original image?

File: torchvision/transforms/transforms.py [hunk 2]
```diff
@@ -1180,3 +1180,78 @@ def __call__(self, img):
...
+                elif isinstance(value, str):
+                    v = torch.rand(img.size()[0], h, w)
+                elif isinstance(value, tuple):
+                    v = torch.FloatTensor(value).view(-1, 1, 1).expand(-1, h, w)
+                return x, y, h, w, v
+
+    def __call__(self, img):
+        """
+        Args:
+            img (Tensor): Image to be erased.
```
[the remark this answers is not in the document]
Comment (rwightman) [reply torchvision/transforms/transforms.py]: If you don't do this operation after Normalization (which is a Tensor based transform), you have to duplicate argument passing and pass your dataset stats to both Normalize and the RandomErasing transform. Post norm, you can assume a mean of 0 and consistent std dev. Also, if you do it before norm, mixed up with other transforms, it's much easier to skew the statistics of your data and cause divergence between train and validation.

In my experience, using it on a few projects now, it's generally cleaner, less fussy, and more efficient (integrated with moving data to the GPU) if done after normalization as tensor ops.

**Comment by rwightman** [comment]
FWIW, I finished several training sessions with different random-erasing settings. I ran no RE, constant (0) RE, and normally distributed (0 mean, 1 std) per-pixel RE. I did not do a random color (solid) run. I'm still running some other tests in this series to validate some other impl and hyper-params for personal interest....

* All training with cosine LR decay, 5 epoch warmup, 120 epochs until a 10 epoch constant cooldown, label smoothing @ 0.1
* Standard imagenet preprocessing & augmentation otherwise
* Thanks to cosine + smoothing, all 3 models I trained are pretty darn good results for a resnet-34 (see torchvision pretrained), including the one without random-erasing
* There is an improvement with each level of RE applied
* The improvement is still evident when validating against the 'ImagenetV2 matched-frequency' validation set, where both RE models show an advantage

ImageNet 1K validation

| training  |  results |
| -- | -- |
| torchvision |  Prec@1 73.312 (26.688) Prec@5 91.426 (8.574) |
| me w/ no RE |  Prec@1 74.772 (25.228) Prec@5 91.988 (8.012) |
| me w/ const 0 RE |  Prec@1 74.838 (25.162) Prec@5 92.212 (7.788) |
| me w/ per-pixel, normal RE | Prec@1 75.110 (24.890) Prec@5 92.284 (7.716) |

ImagenetV2-matched-frequency validation (https://github.com/modestyachts/ImageNetV2)

| training  |  results |
| -- | -- |
| torchvision |  Prec@1 61.190 (38.810) Prec@5 82.710 (17.290) |
| me w/ no RE | Prec@1 62.300 (37.700) Prec@5 83.870 (16.130) |
| me w/ const 0 RE | Prec@1 62.790 (37.210) Prec@5 84.060 (15.940) |
| me w/ per-pixel, normal RE | Prec@1 62.870 (37.130) Prec@5 84.140 (15.860) |

**Comment by zhunzhong07** [comment]
@rwightman Thanks for your experimental results. It is great to see that random erasing could improve the performance of ImageNet.

Did you run these results on your impl https://github.com/rwightman/pytorch-image-models? If so, could you also provide the running shell (i.e., parameters of distributed_train.sh), so that we can accurately reproduce results?

Thank you for your time and efforts on implementing these results.

File: torchvision/transforms/functional.py [hunk 3]
```diff
@@ -796,3 +796,21 @@ def to_grayscale(img, num_output_channels=1):
...
     return img
+
+
+def erase(img, x, y, h, w, value=0):
+    """ Erase the given Image with given value.
+        'Random Erasing Data Augmentation' by Zhong et al.
+        See https://arxiv.org/pdf/1708.04896.pdf
+     Args:
+        img (Tensor): Image to be erased.
+        x: Upper pixel coordinate.
```
[the remark this answers is not in the document]
Comment (zhunzhong07) [reply torchvision/transforms/functional.py]: Thanks. I correct the comments of x and y.

File: torchvision/transforms/functional.py [hunk 4]
```diff
@@ -796,3 +796,21 @@ def to_grayscale(img, num_output_channels=1):
...
+        y: Left pixel coordinate.
```
[the remark this answers is not in the document]
Comment (zhunzhong07) [reply torchvision/transforms/functional.py]: Have corrected it.

**Comment by rwightman** [comment]
Yeah, using the train script in image-models. I was only running single GPU for these runs and did them in parallel. I had a local mod experimenting with the warmup and changing it's overlap behaviour with the main schedule but differences is minor, I extended the epochs here by 5 to compensate. These should reproduce results closely enough:

No RE:
`python train.py /imagenet/ --model resnet34 -b 256 --epochs 125 --warmup-epochs 5 --sched cosine --lr 0.1 --weight-decay 1e-4 --reprob 0.`

RE constant:
`python train.py /imagenet/ --model resnet34 -b 256 --epochs 125 --warmup-epochs 5 --sched cosine --lr 0.1 --weight-decay 1e-4 --reprob 0.4`

RE per-pixel normal:
`python train.py /imagenet/ --model resnet34 -b 256 --epochs 125 --warmup-epochs 5 --sched cosine --lr 0.1 --weight-decay 1e-4 --reprob 0.4 --remode pixel`

**Comment by fmassa** [comment]
@rwightman thanks a lot for the feedback wrt the usefulness of RandomErasing.

I'll have a closer look at the implementation today

**Comment by zhunzhong07** [comment]
@rwightman Thanks! With your provided scripts, I have obtained similar results for ResNet34.

@fmassa Thank you for your attention. I also run RandomErasing for ResNet50 and achieve an improvement (+0.7 in Prec@1).

<p>ImageNet 1K validation</p>
  <table>
  <thead>
  <tr>
  <th>training</th>
  <th>results</th>
  </tr>
  </thead>
  <tbody>
  <tr>
  <td>torchvision ResNet50</td>  
  <td>Prec@1 76.15 (23.85) Prec@5 92.87 (7.13)</td>
  </tr>
  <tr>
  <td>me ResNet50 w/ no RE</td>  
  <td>Prec@1 76.33 (23.67) Prec@5 92.96 (7.04)</td>
  </tr>
  <tr>
  <td>me ResNet50 w/ per-pixel, normal RE</td>  
 <td>Prec@1 77.08 (22.92) Prec@5 93.27 (6.73)</td>
  </tr>
  </tbody>
  </table>

File: torchvision/transforms/functional.py [hunk 5]
```diff
@@ -796,3 +796,20 @@ def to_grayscale(img, num_output_channels=1):
...
+     Args:
+        img (Tensor): Image to be erased.
+        x, y: Upper left pixel coordinate.
+        w: Width of the erased image.
+        h: Height of the erased image.
+        value: Erasing value.
+     Returns:
+        Erased Image.
+    """
+    img[:, x:x + h, y:y + w] = value
```
Comment (fmassa) [inline torchvision/transforms/functional.py]: can you add an assert that `img` is an instance of a `torch.Tensor`?
```python
if not isinstance(img, torch.Tensor):
    raise TypeError("...")
```

File: torchvision/transforms/transforms.py [hunk 6]
```diff
@@ -1180,3 +1180,76 @@ def __call__(self, img):
...
+
+            if w < img.size()[2] and h < img.size()[1]:
+                x = random.randint(0, img.size()[1] - h)
+                y = random.randint(0, img.size()[2] - w)
+                if isinstance(value, numbers.Number):
+                    v = value
+                elif isinstance(value, str):
+                    v = torch.empty([img.size()[0], h, w], dtype=torch.float32).normal_()
+                elif isinstance(value, tuple):
+                    v = torch.FloatTensor(value).view(-1, 1, 1).expand(-1, h, w)
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: Instead of using the legacy `torch.FloatTensor` constructor, can you use instead `torch.tensor(value, dtype=torch.float32)?

File: torchvision/transforms/transforms.py [hunk 7]
```diff
@@ -1180,3 +1180,76 @@ def __call__(self, img):
...
+
+            h = int(round(math.sqrt(target_area * aspect_ratio)))
+            w = int(round(math.sqrt(target_area / aspect_ratio)))
+
+            if w < img.size()[2] and h < img.size()[1]:
+                x = random.randint(0, img.size()[1] - h)
+                y = random.randint(0, img.size()[2] - w)
+                if isinstance(value, numbers.Number):
+                    v = value
+                elif isinstance(value, str):
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: maybe it would be safer to do `isinstance(value, torch._six.string_classes)`?

File: torchvision/transforms/transforms.py [hunk 8]
```diff
@@ -1180,3 +1180,76 @@ def __call__(self, img):
...
+        while True:
+            area = img.size()[1] * img.size()[2]
+
+            target_area = random.uniform(sl, sh) * area
+            aspect_ratio = random.uniform(r1, 1 / r1)
+
+            h = int(round(math.sqrt(target_area * aspect_ratio)))
+            w = int(round(math.sqrt(target_area / aspect_ratio)))
+
+            if w < img.size()[2] and h < img.size()[1]:
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: maybe use `img.shape[2]` and `img.shape[1]` instead in here and below?

File: torchvision/transforms/transforms.py [hunk 9]
```diff
@@ -1180,3 +1180,76 @@ def __call__(self, img):
...
+                x = random.randint(0, img.size()[1] - h)
+                y = random.randint(0, img.size()[2] - w)
+                if isinstance(value, numbers.Number):
+                    v = value
+                elif isinstance(value, str):
+                    v = torch.empty([img.size()[0], h, w], dtype=torch.float32).normal_()
+                elif isinstance(value, tuple):
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: might be safer to do `isinstance(value, (list, tuple))`, if the user passes a `list` instead?

File: torchvision/transforms/functional.py (hunk 5 above)
Comment (zhunzhong07) [reply torchvision/transforms/functional.py]: Thanks. I have added it.

File: torchvision/transforms/transforms.py (hunk 6 above)
Comment (zhunzhong07) [reply torchvision/transforms/transforms.py]: Thanks. I have used instead `torch.tensor(value, dtype=torch.float32).

File: torchvision/transforms/transforms.py (hunk 7 above)
Comment (zhunzhong07) [reply torchvision/transforms/transforms.py]: Thanks. I have used instead 'isinstance(value, torch._six.string_classes)'.

File: torchvision/transforms/transforms.py (hunk 8 above)
Comment (zhunzhong07) [reply torchvision/transforms/transforms.py]: Thanks. I have used img.shape[2] and img.shape[1] instead.

File: torchvision/transforms/transforms.py (hunk 9 above)
Comment (zhunzhong07) [reply torchvision/transforms/transforms.py]: Thanks. I have used instead 'isinstance(value, (list, tuple)) here'.

**Comment by zhunzhong07** [comment]
@fmassa Thank you for your improved comments. I have modified the PR and add an entry to the documentation, according to your suggestions.

Yes, using the mode of `value='random'` would achieve better results. What is your suggestion, only using 'random' mode, or, including multi modes (random and constant)?

**Comment by zhunzhong07** [comment]
Hi @fmassa. I've modified the PR according to your comments. I also add the results of ResNet101 above, consistent improvement is obtained by RandomErasing.

**Comment by Zhaoyi-Yan** [comment]
@zhunzhong07 Hi, have you tried it for detection?

**Comment by zhunzhong07** [comment]
@Zhaoyi-Yan Yes. Random Erasing can improve the results of Fast RCNN on VOC17. Please refer to our [paper.](https://arxiv.org/abs/1708.04896)

**Comment by Zhaoyi-Yan** [comment]
Keeps running in the loop here :
```python
        while True:
            erase_area = random.uniform(scale[0], scale[1]) * area
            aspect_ratio = random.uniform(ratio[0], ratio[1])
            ......
```

File: test/test_transforms.py (hunk 1 above)
Comment (ekagra-ranjan) [reply test/test_transforms.py]: Addressed by #1060.

File: torchvision/transforms/transforms.py [hunk 10]
```diff
@@ -1180,3 +1180,78 @@ def __call__(self, img):
...
+            area = img.size()[1] * img.size()[2]
+
+            target_area = random.uniform(sl, sh) * area
+            aspect_ratio = random.uniform(r1, 1 / r1)
+
+            h = int(round(math.sqrt(target_area * aspect_ratio)))
+            w = int(round(math.sqrt(target_area / aspect_ratio)))
+
+            if w < img.size()[2] and h < img.size()[1]:
+                x = random.randint(0, img.size()[1] - h)
```
[the remark this answers is not in the document]
Comment (ekagra-ranjan) [reply torchvision/transforms/transforms.py]: Addressed by #1087.
